Improve performance of writing FileDrafts

shayneczyzewski opened the pull request:
# Description

This change improves the algorithmic complexity of writing `FileDraft`s by replacing nested list operations (n^2) with `HashMap` based lookup versions (n*logn).

It also fixes a bug that was discovered as part of this process where a `FileDraft` was written twice. It does not add a check to prevent that from reoccurring. That is for another issue: #517 

Fixes #503

## Type of change

Please select the option(s) that is more relevant.

- [x] Code cleanup
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

Martinsos commented on waspc/src/Wasp/Generator/WriteFileDrafts.hs:
```diff
@@ -73,33 +75,34 @@ fileDraftsToWriteAndFilesToDelete Nothing fileDraftsWithChecksums =
 getRedundantGeneratedFiles existingFilePathsToChecksums fileDrafts =
-  let fileDraftPaths = map getDstPath fileDrafts
-   in filter (`notElem` fileDraftPaths) (fst <$> existingFilePathsToChecksums)
+  let fileDraftPathsSet = Set.fromList $ map getDstPath fileDrafts
```
Maybe this would be a bit simpler:
```
let fileDraftPaths = Set.fromList $ getDstPath <$> fileDrafts
 in filter (not . (`member` fileDraftPaths)) (fst <$> existingFilePathsToChecksums)
```
no biggie though, both is fine.

Martinsos reviewed (commented):
One sneaky comment coming in :D

shayneczyzewski commented:
> One sneaky comment coming in :D

![giphy](https://user-images.githubusercontent.com/523636/158482482-3ebe713d-5245-4247-a2d2-414a0642d6cd.gif)
🤣 there will be time for a review lol

While I have you though, I feel like I hit some weird snag where we now get a failure for a single e2e test in CI (server template file does not get processed properly), and locally, but  it seems conditional on `stack clean`... if you build and run on this branch after `stack clean` it fails. Then swap to `main` and run it fails. Then `stack clean` on main again and it passes. Possibly stack related, but I wonder if some race condition in general for Tasty tests occurs once we swap lists to another data structure that may not be threadsafe? Puzzling, but we can use some time tomorrow to chat on if room.

shayneczyzewski commented:
Strange behavior summary (1 of 2):

1) starting on `shayne-martijn-use-maps-and-sets`
- `stack clean`
- `stack test :e2e-test`=> 1 failing test
- `git checkout main`
- `stack test :e2e-test`=> 1 failing test
- `stack clean`
- `stack test :e2e-test`=> ZERO failing tests

shayneczyzewski commented:
Strange behavior summary (2 of 2):

2. starting on `main`
- `stack clean`
- `stack test :e2e-test` => ZERO failing tests
- `git checkout shayne-martijn-use-maps-and-sets`
- `stack test :e2e-test` => ZERO failing tests
- `stack clean`
- `stack test :e2e-test` => 1 failing test

shayneczyzewski commented:
The failing test for reference:
<img width="1830" alt="Screen Shot 2022-03-15 at 3 48 52 PM" src="https://user-images.githubusercontent.com/523636/158494802-a43eec13-7958-4d7c-ae67-c1c8fa419f44.png">

Seems like a template is going unprocessed for some reason.

shayneczyzewski commented:
File: `/Users/shayne/dev/wasp/waspc/e2e-test/test-outputs/waspMigrate-current/waspMigrate/.wasp/out/server/src/server.js`

Martinsos commented:
@shayneczyzewski now I am wondering if it might be the same thing as @sodic had -> if you run `stack test :e2e-test`, but you don't do `stack build` first, you are really running tests for an old build, not for the current version of the code. That is why `stack clean` probably helps, maybe it forces rebuilding? Making this up hm, might not be connected but maybe worth investigating.

Martinsos commented:
@shayneczyzewski so the whole thing with `stack clean`, I think it just forces project to rebuild, so it uses the current version of the code when running tests.

If so, then the problem comes down to: for some reason, on this branch that uses maps, the server.js gets copied uncompiled! And that happens only for `waspMigrate` e2e test.
That is super weird hm.

Martinsos commented:
This is rather fascinating. It is good that we can replicate it consistently, across machines.
On the other hand I have no idea where to start debugging this from.

One thing that changes in this commit is version of strong-path. That is the only thing that is "global", all the other changes seem to be local. They are however affecting writing of file drafts, so some of those local changes could also be the cause. Probalby best to revert piece by piece of code until we detect which one is causing this.

One thing I can think about right now, is that our code for writing template file drafts currently says that if data received is Nothing, it will just copy the template file. And it seems like exactly that is happening. So in case of waspMigrate e2e test, for some reason, no data (`Nothing`) is provided to the template file draft for server.js.

Ok, so let's see what data is supplied to template file draft for server.js .
I just searched for `server.js` in our code, and I see it is generated in ServerGenerator.hs .
And looking at it, I see suspicious part: it is first mentioned in `genSrcDir`, where it is generated with no data supplied, so it is supposed to be just copied, but then it is mentioned again in `genServerJs`! Ha, weird. And `genServerJs` is called in `genSrcDir`, but somewhat later, after the generation of `server.js` without any data.
So, this must be the problem, because we have 2 file drafts for generating server.js -> first without any data (and this one is wrong), and then with data (which is good).
I am guessing the first one, with no data, is old one, when server.js didn't use any templating mechanisms. And then we added mustache stuff into it, and added `genServerJs`, but forgot to remove the old file draft. And that worked fine because the second file draft would overwrite the first one, until for some reason now, that order of writing changed! I guess because we sorted them differently.

Ok great! So this is an issue. We need to remove the old `server.js` draft, so basically remove the 
```
      return $ C.mkSrcTmplFd $ C.asTmplSrcFile [relfile|server.js|],
```
line from `genSrcDir`.

This is an old error that was now exposed.

Now, an interesting question is, how can we avoid this in the future?
Well, first thing is that e2e tests captured this, which is awesome :)! I think this is their biggest feat so far! Without them, we wouldn't capture this, and even if we did figure it out, it would be harder to debug. So they are awesome.
Besides major refactoring of templating system where we get more stuff on the type level, what I think we can do right now is introduce a check that checks if we are attempting to write same file draft twice. That could be present in WriteFileDrafts.hs -> we check if there are two file drafts that have the same destination path. If so, we can throw an error -> it is a bug, it shouldn't be happening, because we don't guarantee the order of writing and it makes everything a mess! How can we even organize them into a map with path being key if we have two things that go under the same key. So yeah, our code there needs to assume there is only one filedraft for each dst path, and it needs to enforce that assumption.

shayneczyzewski commented:
> This is rather fascinating. It is good that we can replicate it consistently, across machines. On the other hand I have no idea where to start debugging this from.
> 
> One thing that changes in this commit is version of strong-path. That is the only thing that is "global", all the other changes seem to be local. They are however affecting writing of file drafts, so some of those local changes could also be the cause. Probalby best to revert piece by piece of code until we detect which one is causing this.
> 
> One thing I can think about right now, is that our code for writing template file drafts currently says that if data received is Nothing, it will just copy the template file. And it seems like exactly that is happening. So in case of waspMigrate e2e test, for some reason, no data (`Nothing`) is provided to the template file draft for server.js.
> 
> Ok, so let's see what data is supplied to template file draft for server.js . I just searched for `server.js` in our code, and I see it is generated in ServerGenerator.hs . And looking at it, I see suspicious part: it is first mentioned in `genSrcDir`, where it is generated with no data supplied, so it is supposed to be just copied, but then it is mentioned again in `genServerJs`! Ha, weird. And `genServerJs` is called in `genSrcDir`, but somewhat later, after the generation of `server.js` without any data. So, this must be the problem, because we have 2 file drafts for generating server.js -> first without any data (and this one is wrong), and then with data (which is good). I am guessing the first one, with no data, is old one, when server.js didn't use any templating mechanisms. And then we added mustache stuff into it, and added `genServerJs`, but forgot to remove the old file draft. And that worked fine because the second file draft would overwrite the first one, until for some reason now, that order of writing changed! I guess because we sorted them differently.
> 
> Ok great! So this is an issue. We need to remove the old `server.js` draft, so basically remove the
> 
> ```
>       return $ C.mkSrcTmplFd $ C.asTmplSrcFile [relfile|server.js|],
> ```
> 
> line from `genSrcDir`.
> 
> This is an old error that was now exposed.
> 
> Now, an interesting question is, how can we avoid this in the future? Well, first thing is that e2e tests captured this, which is awesome :)! I think this is their biggest feat so far! Without them, we wouldn't capture this, and even if we did figure it out, it would be harder to debug. So they are awesome. Besides major refactoring of templating system where we get more stuff on the type level, what I think we can do right now is introduce a check that checks if we are attempting to write same file draft twice. That could be present in WriteFileDrafts.hs -> we check if there are two file drafts that have the same destination path. If so, we can throw an error -> it is a bug, it shouldn't be happening, because we don't guarantee the order of writing and it makes everything a mess! How can we even organize them into a map with path being key if we have two things that go under the same key. So yeah, our code there needs to assume there is only one filedraft for each dst path, and it needs to enforce that assumption.

Amazing debugging @Martinsos that seems to have fixed it! 🥳 Thanks so much. Agree there is value for e2e here in the first real bug, and we can also add extra safeguards to check for duplicate FD paths in this update as well. Thanks again!!!

shayneczyzewski commented on waspc/cabal.project.freeze:
```diff
@@ -58,7 +58,6 @@ constraints: any.Cabal ==3.2.1.0,
              any.conduit ==1.3.4.2,
              any.conduit-extra ==1.3.5,
              any.connection ==0.3.1,
-             any.constraints ==0.13.3,
```
Had a chat on it on hangouts

shayneczyzewski commented on waspc/test/Generator/WriteFileDraftsTest.hs:
```diff
@@ -53,7 +53,9 @@ spec_WriteFileDrafts =
-      fileDraftsToWriteAndFilesToDelete (Just relPathsToChecksums) fdsWithChecksums `shouldBe` (map fst $ newFdsWithChecksums ++ updatedFdsWithChecksums, map fst deletedRelPathsToChecksums)
+      let (fileDraftsToWrite, filesToDelete) = fileDraftsToWriteAndFilesToDelete (Just relPathsToChecksums) fdsWithChecksums
+      fileDraftsToWrite `shouldMatchList` map fst (newFdsWithChecksums ++ updatedFdsWithChecksums)
+      filesToDelete `shouldMatchList` map fst deletedRelPathsToChecksums
```
`shouldMatchList` does a list comparison, but does not care about ordering.

Martinsos replied:
Cool, didn't know about it!

Martinsos reviewed (approved):
LGTM :)!!
